File: Data Structures & Algorithms/mergeSort/submission-0.cs
// Definition for a pair.
// public class Pair {
//     public int Key;
//     public string Value;
//
//     public Pair(int key, string value) {
//         Key = key;
//         Value = value;
//     }
// }
public class Solution {
    public List<Pair> MergeSort(List<Pair> pairs) {
        if(pairs.Count == 0 || pairs.Count == 1) return pairs;

        int mid = pairs.Count/2;
        List<Pair> left = pairs.GetRange(0, mid);
        List<Pair> right = pairs.GetRange(mid, pairs.Count-mid);

        List<Pair> sleft = MergeSort(left);
        List<Pair> sright = MergeSort(right);

        return Merge(sleft, sright);
    }

    private List<Pair> Merge(List<Pair> left, List<Pair> right) {
        int i = 0;
        int j = 0;
        List<Pair> combined = new List<Pair>();
        while(i < left.Count && j < right.Count) {
            if(left[i].Key <= right[j].Key) {
                combined.Add(left[i]);
                i++;
            }
            else {
                combined.Add(right[j]);
                j++;
            }
        }
        if(j != right.Count) {
            while(j!=right.Count) combined.Add(right[j++]); 
        }
        else if(i != left.Count) {
            while(i!=left.Count) combined.Add(left[i++]);
        }
        return combined;
    }
}

File: Data Structures & Algorithms/mergeSort/submission-0.py
# Definition for a pair.
# class Pair:
#     def __init__(self, key: int, value: str):
#         self.key = key
#         self.value = value
class Solution:
    def mergeSort(self, pairs: List[Pair]) -> List[Pair]:
        self.merge_sort(pairs, 0, len(pairs) - 1)
        return pairs

    def merge_sort(self, pairs, start, end):
        if (end - start + 1) <= 1:
            return

        mid = (end + start) // 2
        self.merge_sort(pairs, start, mid)
        self.merge_sort(pairs, mid + 1, end)

        self.merge(pairs, start, mid, end)

    def merge(self, pairs, start, mid, end):
        temp = []

        i = start
        j = mid + 1

        while i <= mid and j <= end:
            if pairs[i].key <= pairs[j].key:
                temp.append(pairs[i])
                i += 1
            else:
                temp.append(pairs[j])
                j += 1

        while i <= mid:
            temp.append(pairs[i])
            i += 1

        while j <= end:
            temp.append(pairs[j])
            j += 1

        for k in range(len(temp)):
            pairs[start + k] = temp[k]
